core/src/main/scala/org/labrad/Connection.scala

```diff
@@ -212,21 +216,69 @@ trait Connection {
   ): Unit = {
     try {
       val loginResponse = credential match {
-        case Password("", password) =>
-          // send first ping packet; response is password challenge
-          val Bytes(challenge) = Await.result(sendManagerRequest(), timeout)
-
-          val md = MessageDigest.getInstance("MD5")
-          md.update(challenge)
-          md.update(UTF_8.encode(CharBuffer.wrap(password)))
-          val data = TreeData("s") // use s instead of y for backwards compatibility
-          data.setBytes(md.digest)
-
-          // send password response; response is welcome message
-          try {
-            Await.result(sendManagerRequest(0, data), timeout)
-          } catch {
-            case e: ExecutionException => throw new IncorrectPasswordException
+        case Password(username @ "", password) =>
+          val useSrp = if (requireSrp) {
+            true
+          } else {
+            // We will use SRP if manager supports it, so we send a PING to get manager features.
+            val resp = Await.result(sendManagerRequest(2, "PING".toData), timeout)
+            val features: Set[String] = resp match {
+              case Cluster(_, Arr(features @ _*)) => Set(features.map(_.getString): _*)
```

Comment:
I'd recommend using a different name for mapping the matched value in `Arr(..)` -- I probably get confused too easily, but seeing the name used twice in immediate parent-child scopes throws me off. It's also not immediately clear what the type of `features` is in the case body. It looks like it may be a Labrad type (an array?), and you're grabbing the elements as a `Seq`. For the sake of readability, could we expand this out a bit without sacrificing performance or functional-ness?